File: DSA09031.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <iterator>

using namespace std;


long long n, k, m;
vector<vector<long long>> matrix(200, vector<long long>(200, 0));

bool check(long long i, long long j) {
    return i >= 0 && j >= 0 && i < 2 * n - 1 && j < 2 * n - 1;
}

void move(long long i, long long j, long long& cnt) {
    static const long long dx[] = {0, 0, -2, 2};
    static const long long dy[] = {2, -2, 0, 0};

    matrix[i][j] = -1;
    for (long long k = 0; k < 4; ++k) {
        long long iNew = i + dx[k];
        long long jNew = j + dy[k];
        if (check(iNew, jNew) && matrix[(i + iNew) / 2][(j + jNew) / 2] == 0) {
            if (matrix[iNew][jNew] == 2) {
                cnt++;
            }
            if (matrix[iNew][jNew] >= 0) {
                move(iNew, jNew, cnt);
            }
        }
    }
}

int main() {
    cin >> n >> k >> m;
    for (long long i = 0; i < m; ++i) {
        long long u, v, x, y;
        cin >> u >> v >> x >> y;
        u = 2 * (u - 1);
        v = 2 * (v - 1);
        x = 2 * (x - 1);
        y = 2 * (y - 1);
        if (u == x) {
            matrix[u][(v + y) / 2] = 1;
        } else {
            matrix[(u + x) / 2][y] = 1;
        }
    }

    for (long long i = 0; i < k; ++i) {
        long long u, v;
        cin >> u >> v;
        u = 2 * (u - 1);
        v = 2 * (v - 1);
        matrix[u][v] = 2;
    }

    vector<long long> ls;
    long long cnt = 1;

    for (long long i = 0; i < 2 * n - 1; i += 2) {
        for (long long j = 0; j < 2 * n - 1; j += 2) {
            if (matrix[i][j] == 2) {
                move(i, j, cnt);
                ls.push_back(cnt);
                cnt = 1;
            }
        }
    }

    long long res = 0;
    for (auto i = 0; i < ls.size(); ++i) {
        for (auto j = i + 1; j < ls.size(); ++j) {
            res += ls[i] * ls[j];
        }
    }


    cout << res << endl;

    return 0;
}

File: DSA09031.py
from itertools import combinations
def function():
    n, k, m = map(int, input().split())

    matrix = [[0] * (2 * n - 1) for i in range(2 * n - 1)]

    for i in range(m):
        u, v, x, y = map(lambda x: 2 * (int(x) - 1), input().split())
        if u == x:
            matrix[u][(v + y) // 2] = "|"
        else:
            matrix[(x + u) // 2][y] = "-"

    for i in range(k):
        ls = [] 
        while not ls:
            ls = list(map(lambda x: int(x) - 1, input().split()))
        matrix[2 * ls[0]][2 * ls[1]] = 2
    
    dx = [0, 0, -2, 2]
    dy = [2, -2, 0, 0]

    ls = []
    cnt = 1

    # for i in matrix:
    #     for j in i:
    #         print("{:}".format(j if j != -1 else "T"), end = " ")
    #     print()
    def check(i, j):
        return 0 <= i < 2 * n - 1 and 0 <= j < 2 * n - 1

    def move(i, j):
        nonlocal cnt
        matrix[i][j] = -1
        for _ in range(4):
            iNew = i + dx[_]
            jNew = j + dy[_]
            if check(iNew, jNew) and matrix[i + dx[_] // 2][j + dy[_] // 2] == 0:
                if matrix[iNew][jNew] == 2:
                    cnt += 1
                if matrix[iNew][jNew] >= 0:
                    move(iNew, jNew)


    for i in range(0, 2 * n - 1, 2):
        for j in range(2 * n - 1):
            if matrix[i][j] == 2:
                move(i, j)
                ls.append(cnt)
                cnt = 1
    res = 0
    for i in combinations(ls, 2):
        res += i[0] * i[1]
    print(res)


if __name__ == "__main__":
    function()
